src/oid4vp/holder.rs

```diff
@@ -0,0 +1,514 @@
+        decoded_request: &AuthorizationRequestObject,
+        request_jwt: String,
+    ) -> anyhow::Result<()> {
+        log::debug!("Verifying DID request.");
+
+        let resolver: VerificationMethodDIDResolver<DIDWeb, AnyJwkMethod> =
+            VerificationMethodDIDResolver::new(DIDWeb);
+
+        verify_with_resolver(
+            &self.metadata,
+            decoded_request,
+            request_jwt,
+            Some(self.trusted_dids.as_slice()),
+            &resolver,
+        )
+        .await?;
+
+        Ok(())
+    }
+}
+
+impl OID4VPWallet for Holder {
+    type HttpClient = oid4vp::core::util::ReqwestClient;
+
+    fn http_client(&self) -> &Self::HttpClient {
+        &self.client
+    }
+
+    fn metadata(&self) -> &WalletMetadata {
+        &self.metadata
+    }
+}
+
+#[async_trait::async_trait]
+impl RequestSigner for Holder {
+    type Error = OID4VPError;
+
+    fn alg(&self) -> Result<String, Self::Error> {
+        Ok(self
+            .jwk()?
+            .algorithm
+            .ok_or(OID4VPError::SigningAlgorithmNotFound(
+                "JWK algorithm not found.".into(),
+            ))?
+            .to_string())
+    }
+
+    fn jwk(&self) -> Result<JWK, Self::Error> {
+        let jwk = self
+            .request_signer
+            .as_ref()
+            .ok_or(OID4VPError::RequestSignerNotFound)?
+            .jwk()?;
+        serde_json::from_str(&jwk).map_err(|e| OID4VPError::JwkParse(e.to_string()))
+    }
+
+    async fn sign(&self, _payload: &[u8]) -> Vec<u8> {
+        tracing::warn!("WARNING: use `try_sign` method instead.");
+
+        Vec::with_capacity(0)
+    }
+
+    async fn try_sign(&self, payload: &[u8]) -> Result<Vec<u8>, Self::Error> {
+        self.request_signer
+            .as_ref()
+            .ok_or(OID4VPError::RequestSignerNotFound)?
+            .try_sign(payload.to_vec())
+            .await
+            .map_err(OID4VPError::from)
+    }
+}
+
+impl JwsSigner for Holder {
+    async fn fetch_info(
+        &self,
+    ) -> Result<ssi::claims::jws::JwsSignerInfo, ssi::claims::SignatureError> {
+        let jwk = self
+            .jwk()
+            .map_err(|e| ssi::claims::SignatureError::Other(e.to_string()))?;
+
+        let algorithm = jwk.algorithm.ok_or(ssi::claims::SignatureError::Other(
+            "JWK algorithm not found.".into(),
+        ))?;
+
+        let key_id = jwk.key_id.clone();
+
+        Ok(ssi::claims::jws::JwsSignerInfo { algorithm, key_id })
+    }
+
+    async fn sign_bytes(
+        &self,
+        signing_bytes: &[u8],
+    ) -> Result<Vec<u8>, ssi::claims::SignatureError> {
+        self.try_sign(signing_bytes)
+            .await
+            .map_err(|e| ssi::claims::SignatureError::Other(format!("Failed to sign bytes: {}", e)))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::oid4vp::request_signer::ExampleRequestSigner;
+    use sd_jwt::SdJwt;
+
+    // NOTE: This test requires the `companion` service to be running and
```

Comment:
@timothee-haudebourg @cobward 

I'm receiving an error while attempting to validate the authorization request url.

```bash
Error: RequestValidation("unable to validate authorization request")
```

I've added this unit test to be run with the companion service to help isolate the problem. I'm going to try to sync with both of you tomorrow morning to further debug.

NOTE: the companion service is based on https://github.com/spruceid/companion/pull/1 which provides an example sd-jwt presentation definition.